[IPV6]: fix BUG of ndisc_send_redirect()
http://www.mail-archive.com/git-commits-head@vger.kernel.org/msg04131.html

     When I tested IPv6 redirect function about kernel 2.6.19.1, and found
    that the kernel can send redirect packets whose target address is global
    address, and the target is not the actual endpoint of communication.
    
      But the criteria conform to RFC2461, the target address defines as
    following:
    
      Target Address An IP address that is a better first hop to use for
                     he ICMP Destination Address.  When the target is
                     the actual endpoint of communication, i.e., the
                     destination is a neighbor, the Target Address field
                     MUST contain the same value as the ICMP Destination
                     Address field.  Otherwise the target is a better
                     first-hop router and the Target Address MUST be the
                     router's link-local address so that hosts can
                     uniquely identify routers.
    
    According to this definition, when a router redirect to a host, the
    target address either the better first-hop router's link-local address
    or the same as the ICMP destination address field. But the function of
    ndisc_send_redirect() in net/ipv6/ndisc.c, does not check the target
    address correctly.
    
    There is another definition about receive Redirect message in RFC2461:
    
    8.1.  Validation of Redirect Messages
    
       A host MUST silently discard any received Redirect message that does
       not satisfy all of the following validity checks:
       ......
       - The ICMP Target Address is either a link-local address (when
         redirected to a router) or the same as the ICMP Destination
         Address (when redirected to the on-link destination).
       ......
    
    And the receive redirect function of ndisc_redirect_rcv() implemented
    this definition, checks the target address correctly.
        if (ipv6_addr_equal(dest, target)) {
            on_link = 1;
        } else if (!(ipv6_addr_type(target) & IPV6_ADDR_LINKLOCAL)) {
            ND_PRINTK2(KERN_WARNING
                   "ICMPv6 Redirect: target address is not link-local.\n");
            return;
        }
    
    So, I think the send redirect function must check the target address
    also.
    
    Signed-off-by: Li Yewang <[EMAIL PROTECTED]>
    Acked-by: YOSHIFUJI Hideaki <[EMAIL PROTECTED]>
    Signed-off-by: David S. Miller <[EMAIL PROTECTED]>

---
From bf0b48dfc368c07c42b5a3a5658c8ee81b4283ac
From: Brian Haley
Date: Mon, 8 Oct 2007 00:12:05 -0700
Subject: [PATCH] [IPv6]: Fix ICMPv6 redirect handling with target multicast address

When the ICMPv6 Target address is multicast, Linux processes the
redirect instead of dropping it.  The problem is in this code in
ndisc_redirect_rcv():

         if (ipv6_addr_equal(dest, target)) {
                 on_link = 1;
         } else if (!(ipv6_addr_type(target) & IPV6_ADDR_LINKLOCAL)) {
                 ND_PRINTK2(KERN_WARNING
                            "ICMPv6 Redirect: target address is not
link-local.\n");
                 return;
         }

This second check will succeed if the Target address is, for example,
FF02::1 because it has link-local scope.  Instead, it should be checking
if it's a unicast link-local address, as stated in RFC 2461/4861 Section
8.1:

       - The ICMP Target Address is either a link-local address (when
         redirected to a router) or the same as the ICMP Destination
         Address (when redirected to the on-link destination).

I know this doesn't explicitly say unicast link-local address, but it's
implied.

This bug is preventing Linux kernels from achieving IPv6 Logo Phase II
certification because of a recent error that was found in the TAHI test
suite - Neighbor Disovery suite test 206 (v6LC.2.3.6_G) had the
multicast address in the Destination field instead of Target field, so
we were passing the test.  This won't be the case anymore.

The patch below fixes this problem, and also fixes ndisc_send_redirect()
to not send an invalid redirect with a multicast address in the Target
field.  I re-ran the TAHI Neighbor Discovery section to make sure Linux
passes all 245 tests now.

Signed-off-by: Brian Haley <brian.haley@hp.com>
Acked-by: David L Stevens <dlstevens@us.ibm.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
---

--- linuz/net/ipv6/ndisc.c	2009-11-07 19:52:20.000000000 +0300
+++ linux/net/ipv6/ndisc.c	2009-11-14 21:43:58.000000000 +0300
@@ -1137,9 +1137,10 @@ static void ndisc_redirect_rcv(struct sk
 
 	if (ipv6_addr_cmp(dest, target) == 0) {
 		on_link = 1;
-	} else if (!(ipv6_addr_type(target) & IPV6_ADDR_LINKLOCAL)) {
+	} else if (ipv6_addr_type(target) !=
+		   (IPV6_ADDR_UNICAST|IPV6_ADDR_LINKLOCAL)) {
 		if (net_ratelimit())
-			printk(KERN_WARNING "ICMP redirect: target address is not linklocal\n");
+			printk(KERN_WARNING "ICMP redirect: target address is not linklocal unicast\n");
 		return;
 	}
 
@@ -1214,6 +1214,12 @@ void ndisc_send_redirect(struct sk_buff
 	if (rt == NULL)
 		return;
 
+	if (!ipv6_addr_cmp(&skb->nh.ipv6h->daddr, target) &&
+	    ipv6_addr_type(target) != (IPV6_ADDR_UNICAST|IPV6_ADDR_LINKLOCAL)) {
+		ND_PRINTK2("ICMPv6 Redirect: target address is not link-local unicast.\n");
+		return;
+	}
+
 	if (rt->rt6i_flags & RTF_GATEWAY) {
 		ND_PRINTK1("ndisc_send_redirect: not a neighbour\n");
 		dst_release(&rt->u.dst);
